File: day16.py
#!/usr/bin/env python3

from get_input import get_input, line_parser
import operator


def filter_aunts(aunts, props):
    for prop, test in props.items():
        new_aunts = {}
        for a, props in aunts.items():
            if prop not in props or test(props[prop]):
                new_aunts[a] = props
        aunts = new_aunts
    return aunts

def part1(aunts, known_pros=None):
    props = {
        'children': lambda v: v == 3,
        'cats': lambda v: v == 7,
        'samoyeds': lambda v: v == 2,
        'pomeranians': lambda v: v == 3,
        'akitas': lambda v: v == 0,
        'vizslas': lambda v: v == 0,
        'goldfish': lambda v: v == 5,
        'trees': lambda v: v == 3,
        'cars': lambda v: v == 2,
        'perfumes': lambda v: v == 1,
    }
    aunts = filter_aunts(aunts, props)
    assert len(aunts) == 1
    return list(aunts.keys())[0]

def part2(aunts, known_pros=None, compare=None):
    props = {
        'children': lambda v: v == 3,
        'cats': lambda v: v > 7,
        'samoyeds': lambda v: v == 2,
        'pomeranians': lambda v: v < 3,
        'akitas': lambda v: v == 0,
        'vizslas': lambda v: v == 0,
        'goldfish': lambda v: v < 5,
        'trees': lambda v: v > 3,
        'cars': lambda v: v == 2,
        'perfumes': lambda v: v == 1,
    }
    aunts = filter_aunts(aunts, props)
    assert len(aunts) == 1
    return list(aunts.keys())[0]

def parse(line):
    aunt_n = line.index(':')
    aunt = int(line[4:aunt_n])
    props = {}
    for prop in line[aunt_n+2:].split(', '):
        k, v = prop.split(': ') 
        props[k] = int(v)
    return (aunt, props)

if __name__ == '__main__':
    aunts = dict(line_parser(get_input(day=16, year=2015), parse=parse))
    # Part 1: 103
    # Part 2: 405
    print("Part 1: {}".format(part1(aunts)))
    print("Part 2: {}".format(part2(aunts)))

File: day16.rs
use common::get_input;
use std::collections::HashMap;
use regex::Regex;

type Input = Vec<HashMap<String, usize>>;

fn part1(sues: &Input) -> usize {
    let facts = HashMap::from([ 
        ("children", 3),
        ("cats", 7),
        ("samoyeds", 2),
        ("pomeranians", 3),
        ("akitas", 0),
        ("vizslas", 0),
        ("goldfish", 5),
        ("trees", 3),
        ("cars", 2),
        ("perfumes", 1),
    ]);
    let matches: Vec<&HashMap<String, usize>> = sues.iter().filter(|sue| sue.iter().all(|(fact, value)| {
        fact == "Sue" ||
        facts[fact.as_str()] == *value
    })).collect();
    assert!(matches.len() == 1);
    matches[0]["Sue"]
}

fn part2(sues: &Input) -> usize {
    let facts = HashMap::from([ 
        ("children", 3),
        ("cats", 7),
        ("samoyeds", 2),
        ("pomeranians", 3),
        ("akitas", 0),
        ("vizslas", 0),
        ("goldfish", 5),
        ("trees", 3),
        ("cars", 2),
        ("perfumes", 1),
    ]);
    let matches: Vec<&HashMap<String, usize>> = sues.iter().filter(|sue| sue.iter().all(|(fact, value)| {
        match fact.as_str() {
            "Sue" => true,
            f if f == "cats" || f == "trees" => facts[f] < *value,
            f if f == "pomeranians" || f == "goldfish" => facts[f] > *value,
            f => facts[f] == *value
        }
    })).collect();
    assert!(matches.len() == 1);
    matches[0]["Sue"]
}

fn parse(text: String) -> Input {
    let re_ader = Regex::new(r"^Sue (\d+): (\w+): (\d+), (\w+): (\d+), (\w+): (\d+)$").unwrap();
    text.trim().split("\n").map(|line| {
        let captures = re_ader.captures(line).unwrap();
        let mut facts: HashMap<String, usize> = (1..4).map(|n| (
                captures.get(2*n).unwrap().as_str().to_string(),
                captures.get(2*n+1).unwrap().as_str().parse::<usize>().unwrap()
        )).collect();
        facts.insert("Sue".to_string(), captures.get(1).unwrap().as_str().parse::<usize>().unwrap());
        facts
    }).collect()
}

fn main() {
    let input = parse(get_input(16, 2015));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
